src/unwrapSteps.ts

```diff
@@ -2,12 +2,25 @@ import { Pipeline, PotentialStep } from '.';
 import { Step } from './base';
 import { Conditional } from './conditional';
 
+function isConditional(x: PotentialStep): x is Conditional<any> {
+    return (
+        'get' in x &&
+        typeof x.get === 'function' &&
+        'accept' in x &&
+        typeof x.accept === 'function'
+    );
+}
+
+class UnknownPotentialStepError extends Error {}
+
 export function unwrapSteps(steps: PotentialStep[]): Step[] {
```

Comment:
I think the `PotentialStep` can be modified for the sake of polymorphism in order to remove the `instanceof` later.
That makes the code structure better as well from my point of view.
```
export type PotentialStep = 
  | { kind: 'pipeline', content: Pipeline }
  | { kind: 'step', content: Step }
  | { kind: 'condition', content: Conditional<Pipeline | Step> };
```

Then it will be
```
export function unwrapSteps(steps: PotentialStep[]): Step[] {
    const ret: Step[] = [];
    for (const s of steps) {
        if (s.kind === 'pipeline') {
            ret.push(...unwrapSteps(s.content.steps));
        } else if (s.kind === 'condition') {
            if (s.content.accept()) {
                const cond = s.content.get();
                if (cond instanceof Pipeline) {
                    ret.push(...unwrapSteps(cond.steps));
                } else {
                    ret.push(cond);
                }
            }
        } else {
            ret.push(s.content);
        }
    }
    return ret;
}
```

Reply:
I think this makes it a lot harder to compose, as you not only have to provide the thing itself (pipeline or step or conditional) but also a wrapper  that does nothing more than helping typescript to infer the right type. It would be a very clunky interface.